src/Pos/Update/Poll/Pure.hs

```diff
@@ -0,0 +1,117 @@
+-- | Pure Poll
+
+module Pos.Update.Poll.Pure
+       ( PurePoll (..)
+       , evalPurePollWithLogger
+       , execPurePollWithLogger
+       , runPurePollWithLogger
+       ) where
+
+import           Universum
+
+import           Control.Lens               (at, to, (%=), (.=))
+import qualified Data.HashMap.Strict        as HM
+import           System.Wlog                (CanLog, HasLoggerName (..), LogEvent,
+                                             NamedPureLogger, logWarning, runNamedPureLog)
+
+import           Pos.Crypto                 (hash)
+import           Pos.Types                  (SoftwareVersion (..))
+import           Pos.Update.Poll.Class      (MonadPoll (..), MonadPollRead (..))
+import qualified Pos.Update.Poll.PollState  as Poll
+import           Pos.Update.Poll.Types      (BlockVersionState (..),
+                                             DecidedProposalState (..),
+                                             UndecidedProposalState (..),
+                                             cpsSoftwareVersion, propStateToEither)
+
+newtype PurePoll a = PurePoll
+    { getPurePoll :: NamedPureLogger (State Poll.PollState) a
+    } deriving (Functor, Applicative, Monad, CanLog, HasLoggerName)
+
+instance HasLoggerName Identity where
+    getLoggerName = mempty
+
+    modifyLoggerName = flip const
+
+runPurePollWithLogger :: Poll.PollState -> PurePoll a -> (a, Poll.PollState, [LogEvent])
+runPurePollWithLogger ps =
+    (\((a, c), b) -> (a, b, c)) . flip runState ps . runNamedPureLog . getPurePoll
+
+evalPurePollWithLogger :: Poll.PollState -> PurePoll a -> a
+evalPurePollWithLogger r = view _1 . runPurePollWithLogger r
+
+execPurePollWithLogger :: Poll.PollState -> PurePoll a -> Poll.PollState
+execPurePollWithLogger r = view _2 . runPurePollWithLogger r
+
+instance MonadPollRead PurePoll where
+    getBVState bv = PurePoll $ use $ Poll.psBlockVersions . at bv
+    getProposedBVs = PurePoll $ use $ Poll.psBlockVersions . to HM.keys
+    getEpochProposers = PurePoll $ use $ Poll.psEpochProposers
+    getConfirmedBVStates = PurePoll $ use $ Poll.psBlockVersions . to HM.toList
+    getAdoptedBVFull = PurePoll $ use $ Poll.psAdoptedBV
+    getLastConfirmedSV an = PurePoll $ use $ Poll.psConfirmedBVs . at an
+    getProposal ui = PurePoll $ use $ Poll.psActiveProposals . at ui
+    getProposalsByApp an = PurePoll $ do
+        delActiveProposalsIndices <- use Poll.psActivePropsIdx
+        activeProposals           <- use Poll.psActiveProposals
+        propGetByApp delActiveProposalsIndices activeProposals
+      where
+        propGetByApp appHashmap upIdHashmap =
+            case HM.lookup an appHashmap of
+                Nothing -> do
+                    logWarning $
+                        "getProposalsByApp: unknown application name " <> pretty an
```

Comment:
@rockbmb wait, you chaned `logWarning` to `logDebug` in a wrong place. It had to be vice versa.